Add sys package

davidcassany opened the pull request:
This commit adds the 'sys' package including interfaces for a shell runner, mounter, logger and syscalls. Most of the code is based on parts of elemental-toolkit/v2 'types' package.

This new 'sys' package provides a global System singleton instance to encompass all this basic utilities. This it is easy at anywhere in the code to access to the configured runner, platform, logger, etc.

Finally it also includes a Makefile target to run the unit tests and compute coverage.

davidcassany commented:
@frelon @fgiudici this is taking some parts of the former elemental-toolkit into an individual and small `sys` package. This aim to include the essentials to interact with the system (fs, runner, syscall, mounter, logger and platform) and defines a global singleton `System` object which can be easily pulled at any time. A clear system method is also provided for tests, I could not figure out a better approach to allow redefining the system in tests.

What you think?

In toolkit having to carryover the Config struct or individual instances all over just to access to the mounter, runner or whatever is something I never liked.

frelon commented:
Based mostly on personal taste I would probably split this into a couple of other packages. An issue I had with the way we did it in the toolkit is that we had these kind of `pkg/types` and `pkg/utils` packages which do not really add much to the developer experience.

A concrete example is I would like to have the platform in it's own package and have the api be something like:

```
import "github.com/suse/elemental/v3/pkg/platform"

...

p := platform.New()

p = platform.Parse("linux/aarch64")
```

I also dislike having singletons in the code, although coming from the toolkit I get what you are trying to solve. Would using the context maybe help with this? Then we could also make use of the background context stuff for cancelling long-running tasks gracefully etc.

davidcassany commented:
> Based mostly on personal taste I would probably split this into a couple of other packages.

No strong opinion on that, all good with me to split it further, having big and loose in purpose packages was annoying in Toolkit, so I am all for it. I am just wondering if we will have to make some packages hierarchy, imagine something like:

```
 pkg/
   sys/
     platform/
     runner/
     syscall/
     logger/
     mounter/
     system.go
```

> I also dislike having singletons in the code, although coming from the toolkit I get what you are trying to solve. Would using the context maybe help with this? Then we could also make use of the background context stuff for cancelling long-running tasks gracefully etc.

That's a really good point, I'll check if we can get rid of this singleton approach by leveraging the context 🤔

fgiudici reviewed (commented):
Grouping everything in a singleton makes it more handy to use and follow, like it!
Waiting to see the version that leverages the Context 😄

davidcassany commented:
@frelon @fgiudici I split the code into multiple packages and adjusted bit the singleton implementation of that global variable. I think having a global variable is better compared to using a context. These are the reasons:
* Having a global context is something that does not really make sense to me (the context should be scoped), then if not global, we need to pass it everywhere as a parameter and that is what I was trying to prevent.
* Using context values is like passing everywhere a `map[interface{}]interface{}`, it is too easy to abuse of it, specially for data that is so generic.
* We are not building a concurrent applications by nature, at most we shell out some commands and run some http get request here and there. So, for cancelling or timeout some operations I´d better implement the context in the specific request, rather than some sort of global background context shared over the whole application. For instance, I can imagine extending the runner to provide async features (e.g. real time logs, show progress) and leverage a context for that.

fgiudici reviewed (approved):
Thanks David, the extra splitting of the packages is nice.
This singleton approach looks like quite an improvement in the usability of the package 💪🏼

frelon commented:
> @frelon @fgiudici I split the code into multiple packages and adjusted bit the singleton implementation of that global variable. I think having a global variable is better compared to using a context. These are the reasons:

Okay, I still think passing the variable into each function avoids the magic of having a global variable/singleton, as well as making the code more testable and reusable. If we really have to go the route of singletons I would argue we should place the packages in `internal/` instead of `pkg/` to signal that we don't want this code to be reused outside of `suse/elemental`

>     * Having a global context is something that does not really make sense to me (the context should be scoped), then if not global, we need to pass it everywhere as a parameter and that is what I was trying to prevent.

Sure, the context should be passed to each function that accepts it. This is pretty much accepted praxis in the go community and resorting to a singleton is in my mind worse than passing around a helper object (context or whatever else).

>     * Using context values is like passing everywhere a `map[interface{}]interface{}`, it is too easy to abuse of it, specially for data that is so generic.

Valid, but you also get alot of value out of context. It can be passed to many functions in the standard library and we can use it for actually testing that error conditions like timeouts work as well as interruptions by for example Ctrl+C.

>     * We are not building a concurrent applications by nature, at most we shell out some commands and run some http get request here and there. So, for cancelling or timeout some operations I´d better implement the context in the specific request, rather than some sort of global background context shared over the whole application. For instance, I can imagine extending the runner to provide async features (e.g. real time logs, show progress) and leverage a context for that.

One does not exclude the other, contexts can be inherited or created anew and often are so we are not excluding those use-cases by starting to leverage context now.

davidcassany commented:
@frelon valid points too. So out of the discussion I think the following:

* Instead of a global sys (or whatever), we can simply explicitly pass it everywhere. Fine with me, at the end this is not much of a big deal.
* We can use context for signals management and handle interruptions, IMHO this is a valid and nice use case.

Leveraging both we can pass the context everywhere and we could eventually embed a system to it. So we make explicit some context is required even if we do not explicitly tell what. We could have some sort of `sys.FromContext(ctx)` method that returns the system variable or panics if not found.

What is unclear to me at this point is how deep this should go, as an example, consider the partitioner object, should the constructor get the context and sys, or only the context and assume there will be a sys in it?

fgiudici reviewed (approved):
LGTM

atanasdinov commented on pkg/sys/system.go:
```diff
@@ -0,0 +1,119 @@
+	if sysObj.Platform == nil {
+		defaultPlatform, err := platform.NewPlatformFromArch(runtime.GOARCH)
+		if err != nil {
+			sysObj.Logger.Errorf("error parsing default platform (%s): %s", runtime.GOARCH, err.Error())
```
Isn't the caller supposed to drop this?

atanasdinov commented on pkg/sys/system.go:
```diff
@@ -0,0 +1,119 @@
+	Platform *platform.Platform
+}
+
+type SystemOpts func(a *System) error
```
Could you please explain why exactly we need these if all fields are exported in the `System` type? AFAICS, all the option functions are used in the tests in the current elemental-toolkit base, however, we should be able to just populate them directly too.

This way outside of tests we can just call `NewSystem()` and drop the more complex structures e.g. only set a logger for the runner part, etc.

davidcassany replied:
Is mostly based on personal preference of not flooding a constructor with tons of parameters, so this way they are options. I do not consider an acceptable pattern calling the constructor and right after having to adjust the default values. I know all types are exported there, I have also considered not exporting them and simply create accessor methods, so make explicit to the caller these values should not be altered. I would also be fine with that approach.

atanasdinov commented on pkg/sys/mock/runner.go:
```diff
@@ -0,0 +1,158 @@
+	"github.com/suse/elemental/v3/pkg/sys/log"
+)
+
+type FakeRunner struct {
```
Can we drop all "Test" and "Fake" prefixes from the type names? They're now all coming from `mock` so it is clear they're such.

atanasdinov commented on pkg/sys/vfs/fs.go:
```diff
@@ -0,0 +1,333 @@
+type FS interface {
+	Chmod(name string, mode fs.FileMode) error
+	Create(name string) (*os.File, error)
+	Glob(pattern string) ([]string, error)
```
This doesn't seem used in elemental-toolkit. Can we drop it?

atanasdinov commented on pkg/sys/vfs/fs.go:
```diff
@@ -0,0 +1,333 @@
+}
+
+// permError returns an *os.PathError with Err syscall.EPERM.
+func permError(op, path string) error {
```
This seems like it's not going to be used. Can we remove it?

atanasdinov commented on pkg/sys/vfs/fs.go:
```diff
@@ -0,0 +1,333 @@
+	return vfs.OSFS
+}
+
+func ReadOnlyFS() FS {
```
Is this supposed to be used only in tests?

davidcassany replied:
I guess we can drop it. We have some specific one for tests.

atanasdinov commented on pkg/sys/vfs/fs.go:
```diff
@@ -0,0 +1,333 @@
+	TempDirPerm    = os.ModePerm | os.ModeSticky | os.ModeDir
+)
+
+type FS interface {
```
Can we move it to where it is going to be used?

davidcassany replied:
This is a tricky one, we can move it sys together with all the non typed functions, but then we will have to also implement some functions. Let me give it a try.

atanasdinov replied:
Oh, it is because the interface is actually the first argument to many functions and is technically used here.

atanasdinov replied:
Yeah, this is a little hard. I guess let's keep it as is for now and we can see whether we can improve it later. Sounds like a plan?

davidcassany replied:
Already moved it to system package now.

davidcassany commented on pkg/sys/log/logger.go:
```diff
@@ -0,0 +1,169 @@
+limitations under the License.
+*/
+
+package log
```
No strong opinion in that regard, I mostly added here because I have included it in `System` structure. In my head the System object is what will be used all over the project and in encompasses few interfaces to interact with the system (including logging) and the target platform.

davidcassany commented on pkg/sys/log/logger.go:
```diff
@@ -0,0 +1,169 @@
+)
+
+// Logger is the interface we want for our logger, so we can plug different ones easily
+type Logger interface {
```
Sure, I just copied over what was there, we can limit the scope of this logging interface.

davidcassany commented on pkg/sys/log/logger_test.go:
```diff
@@ -0,0 +1,64 @@
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus" // TODO it should not be required, bad isolation
```
yes this is an old legacy comment there

davidcassany commented on pkg/sys/platform/platform.go:
```diff
@@ -0,0 +1,154 @@
+	Archx86     = "x86_64"
+	ArchArm64   = "arm64"
+	ArchAarch64 = "aarch64"
+	ArchRiscV64 = "riscv64"
```
Yes we can

davidcassany commented on pkg/sys/platform/platform.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func (p *Platform) CustomUnmarshal(data interface{}) (bool, error) {
```
Yes this can be drop. I don't expect to use mapstructure

davidcassany commented on pkg/sys/platform/platform.go:
```diff
@@ -0,0 +1,154 @@
+	return fmt.Sprintf("%s/%s", p.OS, p.GolangArch)
+}
+
+func (p Platform) MarshalYAML() (interface{}, error) {
+	return p.String(), nil
+}
+
+func (p *Platform) UnmarshalYAML(value *yaml.Node) error {
```
We can drop, will be easy to add them if needed

davidcassany commented on pkg/sys/mounter/mounter.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+func NewMounter(binary string) Mounter {
+	return mount.NewWithoutSystemd(binary)
```
It is there, it is in the import, we are just not touching anything from the interface we are based, we just use few methods as they are. So it is easy to switch to another library if needed.

atanasdinov replied:
Actually it doesn't seem to be there:
```
» go build ./...                                                                                             [13:17:36]
# github.com/suse/elemental/v3/pkg/sys/mounter
pkg/sys/mounter/mounter.go:28:15: undefined: mount.NewWithoutSystemd
```

Perhaps it's in another module?

davidcassany commented on pkg/sys/mounter/mounter.go:
```diff
@@ -0,0 +1,44 @@
+	//GetMountRefs(path string) ([]string, error)
+}
+
+func NewMounter(binary string) Mounter {
```
This function returns the interface instance, the binary path is just used to initiate the interface, we can assume it will be always constant, but I do not really see a benefit out of it, just constraints.

atanasdinov replied:
Yes, sorry. The point was to use the binary path within the function instead of having it exposed to the caller i.e. `mounter.NewMounter()` instead of `mounter.NewMounter(mounter.Binary)`.

davidcassany commented on pkg/sys/mounter/mounter.go:
```diff
@@ -0,0 +1,44 @@
+	return mount.NewWithoutSystemd(binary)
+}
+
+func NewDummyMounter() Mounter {
```
We can drop it and re-add if needed. This not for testing, this is for running unprivileged tasks that required a full System type (in former Toolkit an overloaded Config type).

davidcassany commented on pkg/sys/runner/runner.go:
```diff
@@ -0,0 +1,76 @@
+	Run(string, ...string) ([]byte, error)
+	RunCmd(cmd *exec.Cmd) ([]byte, error)
+	CommandExists(command string) bool
+	GetLogger() log.Logger
+	SetLogger(logger log.Logger)
```
Fine with me

davidcassany commented on pkg/sys/log/logger.go:
```diff
@@ -0,0 +1,169 @@
+var emojiStrip = regexp.MustCompile(`[:][\w]+[:]`)
+
+func (w *logrusWrapper) Debug(args ...any) {
+	converted := convert(args)
```
No strong opinion, I am just fine by making this simpler and just keep one of the two methods. Moreover I think we can consider adopting a different logging library too, I just saw logrus is in maintenance mode. So probably we should narrow the interface so it is easier switch to another library.

davidcassany replied:
I'd suggest to stick with something like `Debug(msg string, args ...any)`, so we are setting an interface that matches the methods from `log/slog` from the native library.

davidcassany replied:
Removed the dual `Debug` vs `Debugf`, and only left `Debug(string, ...any)`.

atanasdinov commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,131 @@
+	GetLevel() uint32
+	SetOutput(writer io.Writer)
+
+	SetContext(string)
+	SpinnerStop()
+	Spinner()
```
Can we safely drop these? They aren't used in `elemental-toolkit` either.

atanasdinov commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,131 @@
+	Debug(string, ...any)
+	Error(string, ...any)
+	Fatal(string, ...any)
+	Warning(string, ...any)
```
I guess we can also drop this one as we already have `Warn()`?

atanasdinov commented on pkg/log/logger_test.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+var _ = Describe("logger", Label("log"), func() {
+	It("TestNewLogger returns a logger interface", func() {
```
The descriptions now have to be adjusted to no longer reflect the old function names.

atanasdinov commented on pkg/sys/vfs/fs.go:
```diff
@@ -0,0 +1,100 @@
+	osfs vfs.FS
+}
+
+func OSFS() *vfsOS {
```
This highlights the following warning / linter error:
> Exported function with the unexported return type 

Also applies to the other types we initialise.

davidcassany replied:
Yes, I prefer to keep interface implementations unexported, this way we can keep implementation and interface entirely separate and we enforce the interface use everywhere else.

I don't see any problem here.

atanasdinov reviewed (commented):
Thanks for addressing the comments! Few final ones and I believe this is good to go!

atanasdinov reviewed (approved):
Thanks for addressing all the comments -- as mentioned offline I believe we're off to a great start! 🚀 

I'm not fully convinced on keeping the implementation types unexported as I don't see how those being exported collides with interface usages, but let's get going and we can re-evaluate later once we have more code in. I'm happy to help to improve the structure later on too! 

Great work!

davidcassany commented:
> I'm not fully convinced on keeping the implementation types unexported as I don't see how those being exported collides with interface usages, but let's get going and we can re-evaluate later once we have more code in. I'm happy to help to improve the structure later on too!

Sure, no strong opinion on that either. This is mostly my personal preference of not exposing the specific implementations, so to ensure we are always accessing the instance through the interface. I am opened to change that though.

davidcassany commented:
Rebased on a single commit before merging
